File: swift/NucleotideCount.swift
class DNA {
    let strand: String
    let symbols = "ACGT"
    
    func counts() -> [String: Int] {
        var dict: [String: Int] = [:]

        symbols.forEach { dict[String($0)] = count(String($0)) }
        
        return dict
    }
    
    func count(_ ch: String) -> Int {
        return strand.filter { String($0) == ch }.count
    }
    
    init?(strand s: String) {
        for symbol in s {
            if !symbols.contains(symbol) { return nil }
        }
        
        strand = s
    }
}

File: clojure/nucleotide_count.clj
(ns nucleotide-count)

(defn count
  "Returns the amount of times `ch` appears in `s`."
  [ch s]
  (when-not (.contains [\A \T \C \G] ch) (throw IllegalArgumentException))
  (clojure.core/count (filter #(= ch %) s)))

(defn nucleotide-counts
  "Returns occurences of each nucleotide in a given DNA string."
  [s]
  (let [keys [\A \T \C \G]]
    (zipmap keys (map #(count % s) keys))))
